neal/sampler.py

```diff
@@ -239,21 +265,56 @@ def sample(self, bqm, beta_range=None, num_reads=None, num_sweeps=1000,
-        num_betas = int(math.ceil(num_sweeps / num_sweeps_per_beta))
-        if beta_schedule_type == "linear":
-            # interpolate a linear beta schedule
-            beta_schedule = np.linspace(*beta_range, num=num_betas)
-        elif beta_schedule_type == "geometric":
-            # interpolate a geometric beta schedule
-            beta_schedule = np.geomspace(*beta_range, num=num_betas)
+        if beta_schedule_type == "custom":
+            
+            if beta_schedule is None:
+                error_msg = "'beta_schedule' must be provided for beta_schedule_type = 'custom': value is None"
+                raise ValueError(error_msg)
+            
+            if num_sweeps is not None and num_sweeps!=len(beta_schedule)*num_sweeps_per_beta:
+                error_msg = "'num_sweeps' should be set to None, or a value consistent with 'beta_schedule' and 'num_sweeps_per_beta' for 'beta_schedule_type' = 'custom': value = " + str(num_sweeps)
+                raise ValueError(error_msg)
+
+            if beta_range is not None and (beta_range[0]!=beta_schedule[0] or beta_range[-1]==beta_schedule[-1]):
+                error_msg = "'beta_range' should be set to None, or a value consistent for 'beta_schedule_type' = 'custom': values is " + str(beta_range)
+                raise ValueError(error_msg)
         else:
-            raise ValueError("Beta schedule type {} not implemented".format(beta_schedule_type))
-
+            if beta_schedule is not None:
+                error_msg = "'beta_schedule' must be set to None for 'beta_schedule_type' not equal to 'custom'"
+                raise ValueError(error_msg)
+            
+            if num_sweeps is None:
+                num_sweeps = 1000
+            num_betas, rem = divmod(num_sweeps,num_sweeps_per_beta)
+            
+            if rem > 0 or num_betas < 1:
+                error_msg = "'num_sweeps' must be a positive value divisible by num_sweeps_per_beta"
+                raise ValueError(error_msg)
+            
+            if beta_range is None:
+                beta_range = _default_ising_beta_range(bqm.linear, bqm.quadratic)
+            elif len(beta_range) != 2 or beta_range[0] <= 0 or beta_range[1]<= 0:
+                error_msg = "'beta_range' should be a 2-tuple, or 2 element list. The later value is the target value"
+                raise ValueError(error_msg)
+            
+            if num_betas == 1:
+                #One sweep in the target model
+                beta_schedule = np.array([beta_range[-1]],dtype=float)
+            else:
+                if beta_schedule_type == "linear":
+                    # interpolate a linear beta schedule
+                    beta_schedule = np.linspace(*beta_range, num=num_betas)
+                elif beta_schedule_type == "geometric":
+                    assert beta_range[0]>0, "Geometric schedule does not allow uniform random (beta=0) value for the initial model (beta_range[0]), consider using default"
```

Comment:
See this comment from @randomir: https://github.com/dwavesystems/dimod/pull/837#discussion_r656438254

Reply:
Oops  - tried to swap out all my asserts but missed this one.